File: src/main/java/org/micoli/php/service/PsiElementUtil.java
package org.micoli.php.service;

import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.Computable;
import com.intellij.openapi.util.TextRange;
import com.intellij.psi.PsiDocumentManager;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.PsiNamedElement;
import com.intellij.usageView.UsageInfo;
import org.jetbrains.annotations.NotNull;
import org.micoli.php.ui.popup.FileExtract;

public class PsiElementUtil {
    public static PsiElement findFirstLeafElement(PsiElement element) {
        PsiElement children = element.getFirstChild();
        if (children == null) {
            return element;
        }
        return findFirstLeafElement(children);
    }

    public static FileExtract getFileExtract(PsiElement element, int lineCount) {
        return getFileExtract(element.getProject(), element.getContainingFile(), element.getTextRange(), lineCount);
    }

    public static FileExtract getFileExtract(UsageInfo usageInfo, int lineCount) {
        return getFileExtract(usageInfo.getProject(), usageInfo.getFile(), usageInfo.getRangeInElement(), lineCount);
    }

    public static FileExtract getFileExtract(Project project, PsiFile file, TextRange range, int lineCount) {
        Document document = PsiDocumentManager.getInstance(project).getDocument(file);

        if (document != null) {
            int lineNumber = document.getLineNumber(range.getStartOffset());

            return new FileExtract(
                    lineNumber + 1,
                    document.getText(TextRange.create(
                            document.getLineStartOffset(lineNumber),
                            document.getLineEndOffset(lineNumber + lineCount))));
        }

        return new FileExtract(-1, "");
    }

    public static @NotNull String getHumanReadableElementLink(PsiElement element) {

        return ApplicationManager.getApplication().runReadAction((Computable<String>) () -> {
            String base = PathUtil.getPathWithParent(element.getContainingFile(), 2);
            String format = "%s#%s: %s";

            if (element instanceof PsiNamedElement) {
                return String.format(
                        format, base, getFileExtract(element, 1).lineNumber(), ((PsiNamedElement) element).getName());
            }

            return String.format(
                    format,
                    base,
                    getFileExtract(element, 1).lineNumber(),
                    element.getText().replaceAll("\\s( *)", " "));
        });
    }
}

File: src/main/kotlin/org/micoli/php/service/intellij/psi/PsiElementUtil.kt
package org.micoli.php.service.intellij.psi

import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.project.Project
import com.intellij.openapi.util.Computable
import com.intellij.openapi.util.TextRange
import com.intellij.psi.PsiDocumentManager
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import com.intellij.psi.PsiNamedElement
import com.intellij.usageView.UsageInfo
import org.micoli.php.service.filesystem.PathUtil
import org.micoli.php.ui.popup.FileExtract

object PsiElementUtil {
    @JvmStatic
    fun findFirstLeafElement(element: PsiElement): PsiElement {
        val children = element.firstChild ?: return element
        return findFirstLeafElement(children)
    }

    @JvmStatic
    fun getFileExtract(element: PsiElement, lineCount: Int): FileExtract {
        return getFileExtract(element.project, element.containingFile, element.textRange, lineCount)
    }

    @JvmStatic
    fun getFileExtract(usageInfo: UsageInfo, lineCount: Int): FileExtract {
        return getFileExtract(
            usageInfo.project, usageInfo.file!!, usageInfo.rangeInElement!!, lineCount)
    }

    @JvmStatic
    fun getFileExtract(
        project: Project,
        file: PsiFile,
        range: TextRange,
        lineCount: Int
    ): FileExtract {
        val document = PsiDocumentManager.getInstance(project).getDocument(file)

        if (document != null) {
            val lineNumber = document.getLineNumber(range.startOffset)

            return FileExtract(
                PathUtil.getPathWithParent(file, 2),
                lineNumber + 1,
                document.getText(
                    TextRange.create(
                        document.getLineStartOffset(lineNumber),
                        document.getLineEndOffset(lineNumber + lineCount))))
        }

        return FileExtract(null, -1, "")
    }

    @JvmStatic
    fun getHumanReadableElementLink(element: PsiElement): String {
        return ApplicationManager.getApplication()
            .runReadAction(
                Computable {
                    val base = PathUtil.getPathWithParent(element.containingFile, 2)
                    val format = "%s#%s: %s"

                    if (element is PsiNamedElement) {
                        return@Computable String.format(
                            format, base, getFileExtract(element, 1).lineNumber, element.name)
                    }
                    String.format(
                        format,
                        base,
                        getFileExtract(element, 1).lineNumber,
                        element.text.replace("\\s( *)".toRegex(), " "))
                })
    }
}
